flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -804,22 +808,31 @@ private boolean isFulfillableByRegisteredSlots(ResourceProfile resourceProfile)
 	}
 
 	private Optional<PendingTaskManagerSlot> allocateResource(ResourceProfile resourceProfile) {
-		final Collection<ResourceProfile> requestedSlots = resourceActions.allocateResource(resourceProfile);
+		if (workerResourceSpec == null) {
+			// standalone mode, cannot allocate resource
```

Comment:
I am also curious about why do we make standalone cluster so different? Before this PR, the `StandaloneResourceManager#startNewWorker` is also called and just return an empty collection.

Reply:
The problem is not about calling `StandaloneResourceManager#startNewWorker`. It is about not having a `WorkerResourceSpec` on standalone setups.

For a standalone cluster, use may have TM resource configurations only on the TM machines. That means trying to create a `WorkerResourceSpec` from the configuration on JM might lead to failures due to missing of memory configurations. I think it does not make sense to force the users to set TM configurations on the JM machine, because they are not really used.

We have run into such problems during the 1.9 release testing, where a standalone JM failed because it tries to calculate a default TM managed memory size.

Reply:
It should be able to merge into `if (!resourceActions.allocateResource(workerResourceSpec))` branch. But I am not very sure because the next condition is

```
if (!Preconditions.checkNotNull(slotResourceProfile,
			"slotResourceProfile should be null iff workerResourceSpec is null, which means standalone mode.")
			.isMatching(resourceProfile))
```

where we now have three profile(including workerResourceSpec) things become quite complex.

Reply:
@xintongsong  So where is a document mentions what's the different among `slotResourceProfile`, `resourceProfile` and  `workerResourceSpec`. I can see there are some constraint among them but it is hard to understand from my side.

Reply:
@TisonKun
Please see my reply to Yang for why calling `allocateResource` on standalone setup. For the same reason, we cannot merge this into the `if (!resourceActions.allocateResource(workerResourceSpec))` branch.

Regarding `slotResourceProfile`, `resourceProfile` and `workerResourceSpec`.
- `slotResourceProfile` describes the resource capacity of a slot.
- `resourceProfile` describes the resource requirement of a slot request.
- `workerResourceSpec` describes the resource requirement of a worker (TM) to request.
I think you're right, these things are indeed hard to understand. How do you think about renaming `resourceProfile` to `requestedSlotResourceProfile`, and `slotResourceProfile` to `defaultSlotResourceProfile`?

Reply:
@xintongsong thanks for your explanation! That sounds good.

Reply:
Yes, you are right. We should not require the users to set TM resource related configurations on JM side for standalone. This is the right direction. 
What i am thinking is when there are some future changes of the `SlotManagerImpl`, we always need to be careful and keep in mind that `WorkerResourceSpec` should be `null` for standalone cluster.

Reply:
@wangyang0918
True. That's why we annotate it as `@Nullable` and, as suggested by Tison and updated to the PR, clearly documented that a `null` value means in standalone mode.